edge/pkg/devicetwin/dtmanager/twin_test.go

```diff
@@ -0,0 +1,1604 @@
+
+	"github.com/astaxie/beego/orm"
+	"github.com/golang/mock/gomock"
+	"github.com/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/edge/mocks/beego"
+	"github.com/kubeedge/kubeedge/edge/pkg/common/dbm"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtclient"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcontext"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dttype"
+)
+
+var ormerMock1 *beego.MockOrmer
+var querySeterMock1 *beego.MockQuerySeter
+
+func initMocks1(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	ormerMock1 = beego.NewMockOrmer(mockCtrl)
+	querySeterMock1 = beego.NewMockQuerySeter(mockCtrl)
+	dbm.DBAccess = ormerMock1
+}
+
+func (tw TwinWorker) sendMsg(msg *dttype.DTMessage, msgheart string, actionType string, contentType interface{}) {
+	if tw.ReceiverChan != nil {
+		msg.Action = actionType
+		msg.Msg.Content = contentType
+		tw.ReceiverChan <- msg
+	}
+	if tw.HeartBeatChan != nil {
+		tw.HeartBeatChan <- msgheart
+	}
+}
+
+func recMsg(channel chan interface{}, message *dttype.DTMessage) {
+	msg, _ := <-channel
+	*message = *msg.(*dttype.DTMessage)
+	return
+}
+
+func expectedFunc() *dttype.TwinValue {
+	var expected dttype.TwinValue
+	value := "value"
+	valueMetaData := &dttype.ValueMetadata{Timestamp: time.Now().UnixNano() / 1e6}
+	expected.Value = &value
+	expected.Metadata = valueMetaData
+	return &expected
+}
+
+func actualFunc() *dttype.TwinValue {
+	var actual dttype.TwinValue
+	value := "value"
+	valueMetaData := &dttype.ValueMetadata{Timestamp: time.Now().UnixNano() / 1e6}
+	actual.Value = &value
+	actual.Metadata = valueMetaData
+	return &actual
+}
+
+func keyTwinUpdateFunc() dttype.DeviceTwinUpdate {
+	var keyTwinUpdate dttype.DeviceTwinUpdate
+	twinKey := make(map[string]*dttype.MsgTwin)
+	twinKey["key1"] = &dttype.MsgTwin{
+		Expected: expectedFunc(),
+		Actual:   actualFunc(),
+		Metadata: &dttype.TypeMetadata{"nil"},
+	}
+	keyTwinUpdate.Twin = twinKey
+	keyTwinUpdate.BaseMessage = dttype.BaseMessage{EventID: "Event1"}
+	return keyTwinUpdate
+}
+
+func twinWorkerFunc(receiverChannel chan interface{}, confirmChannel chan interface{}, heartBeatChannel chan interface{}, context dtcontext.DTContext, group string) TwinWorker {
+	return TwinWorker{
+		Worker{
+			receiverChannel,
+			confirmChannel,
+			heartBeatChannel,
+			&context,
+		},
+		group,
+	}
+}
+
+func contextFunc() dtcontext.DTContext {
+	context := dtcontext.DTContext{
+		DeviceList:  &sync.Map{},
+		DeviceMutex: &sync.Map{},
+		Mutex:       &sync.Mutex{},
+	}
+	var testMutex sync.Mutex
+	context.DeviceMutex.Store("DeviceB", &testMutex)
+	var device dttype.Device
+	context.DeviceList.Store("DeviceB", &device)
+	return context
+}
+
+func msgFunc(content interface{}) interface{} {
+	return &model.Message{
+		model.MessageHeader{},
+		model.MessageRoute{},
+		content,
+	}
+}
+func msgFunc1(content interface{}) interface{} {
```

Comment:
what is the difference btwn msgFunc and msgFunc1 ??

Reply:
One function returns the address to model.Message object and the other function returns model.Message object